src/cell.cpp

```diff
@@ -768,50 +773,88 @@ bool CSGCell::contains_simple(Position r, Direction u, int32_t on_surface) const
-      // If the token is not an operator, evaluate the sense of particle with
-      // respect to the surface and see if the token matches the sense. If the
-      // particle's surface attribute is set and matches the token, that
-      // overrides the determination based on sense().
-      i_stack++;
+  // For each token in prefix
+  for (auto it = region_prefix_.begin(); it != region_prefix_.end(); it++) {
+    // Dereference current iterator
+    int32_t token = *it;
+
+    // If the token is a surface
+    if (token < OP_UNION) {
+      // Evaluate the sense of the particle with respect to the surface and see
+      // if the token matches the sense. If the particle's surface attribute is
+      // set and matches the token, that overrides the determination based on
+      // sense.
       if (token == on_surface) {
-        stack[i_stack] = true;
+        in_cell = true;
       } else if (-token == on_surface) {
-        stack[i_stack] = false;
+        in_cell = false;
       } else {
         // Note the off-by-one indexing
         bool sense = model::surfaces[abs(token) - 1]->sense(r, u);
-        stack[i_stack] = (sense == (token > 0));
+        in_cell = (sense == (token > 0));
       }
+
+      // While the operator stack is populated
+      while (op_stack.size() > 0) {
+        // Extract the top operator
+        int32_t op = op_stack.back();
+        op_stack.pop_back();
+
+        // If the operator is a union and the particle is in the cell or the
+        // the operator is an intersection and the particle is not in the cell
+        if ((op == OP_UNION && in_cell == true) ||
+            (op == OP_INTERSECTION && in_cell == false)) {
+          // Initialize counters for proceeding surfaces and operators
+          int number_of_operators = 0;
+          int number_of_surfaces = 0;
+
+          // While the counter for surfaces is not one more than the
+          // number of operators
+          while (number_of_surfaces != number_of_operators + 1) {
+            // Move iterator forward and fetch next token
+            it++;
+            int32_t next_token = *it;
+
+            // If the next token is a surface add to the surface counter, 
+            // if the next token is either an intersection or a union
+            // add to the operator counter
+            if (next_token < OP_UNION) {
+              number_of_surfaces++;
+            } else if (next_token < OP_COMPLEMENT) {
+              number_of_operators++;
+            }
+          }
+
+        // If the operator is a complement
+        } else if (op == OP_COMPLEMENT) {
+          // Set the current cell boolean to the opposite
+          in_cell = !in_cell;
+
+        // If none of the above are true, break
+        } else {
+          break;
+        }
+
+        // If one of the above was true and the size of the operator
+        // stack is zero return the current cell boolean
+        if (op_stack.size() == 0) {
+          return in_cell;
+        }
```

Comment:
With our short-circuit skipping, if the operator stack is empty, this means that we've reached the end of the expression, so I believe we can omit this entirely (I tried doing that on the ITER model I'm working with and it seems to work -- check for yourself though)
```suggestion
```